system/src/system_control_internal.cpp

```diff
@@ -389,6 +415,44 @@ void system_ctrl_set_result(ctrl_request* req, int result, ctrl_completion_handl
     particle::system::SystemControl::instance()->setResult(req, result, handler, data);
 }
 
+int system_ctrl_add_request_filter(system_control_acl default_action, system_control_filter* filters, void* reserved) {
+    if (default_action == SYSTEM_CONTROL_ACL_NONE) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+    particle::system::SystemControl::instance()->filterCtrlReq_.vecCtrlReq_.clear();
+
+    // If default action is to accept requests, vecCtrlReq_ will contain requests that need to be rejected
+    if (default_action == SYSTEM_CONTROL_ACL_ACCEPT) {
```

Comment:
No need for such complexity. To simplify just pass default action and `filters` as-is to the class and set two internal member variables: `defaultAction_` and keep `filters` struct as-is. When filtering check the default action and traverse the list, no need for an extra Vector. When replacing the list of filters make sure to free the memory of the list members.